### PR TITLE
Dev api tests

### PR DESCRIPTION
**Pull request by FrauDeborah**
Johnathan,

I closed the qrlApiTests and created another branch devApiTests which has all the current dev updates in the branch.

Basically, this PR has updated tests for the api service, ng lint was run with the recommended lint corrections made.

Deborah

### CODE REVIEW

File: src/app/_services/api.service.spec.ts [hunk 1]
```diff
@@ -6,13 +6,20 @@ import { HttpClientTestingModule, HttpTestingController} from '@angular/common/h
...
 import { ApiService } from './api.service';
+import { environment } from './../../environments/environment';
 
 describe('ApiService', () => {
 
   let service: ApiService;
   let httpClient: HttpClient;
   let httpTestingController: HttpTestingController;
 
+  const testFackUrl = '/api/testFackUrl';
```
Comment (haxwell) [inline src/app/_services/api.service.spec.ts]: Do you mean `fake` (instead of `fack`)?

File: src/app/_services/api.service.spec.ts [hunk 2]
```diff
@@ -22,26 +29,159 @@ describe('ApiService', () => {
...
+  it('#api.get should be defined ', () => {
+    expect(service.get(testFackUrl, testUsername, testPassword)).toBeDefined();
+  });
+
+  it('#api.get can test HttpClient.get', () => {
+  //  const testFackData = {name: 'Test Fack Data'};
+    const testHeader = service.getHttpHeadersWithBase64Authorization(testUsername, testPassword);
+    // Make an HTTP GET request
+    httpClient.get<any>(testBaseFackUrl, testHeader)
+      .subscribe(data =>
```
Comment (haxwell) [inline src/app/_services/api.service.spec.ts]: The comments I made on #30 re: unit tests, apply here too. Basically, your tests are currently testing Angular's `HTTPClient`. We need to test `api.service`.

**Comment by FrauDeborah** [comment]
Johnathan (@haxwell),

Needing your advise on the **devApiTests** regarding the failing `tests apiService.get()`.

I removed all the previous tests and added the `tests apiService.get()` using `url = /api/verifyCredentials`. I managed to determine and fix most of the errors, but I don't understand the following error. 

image

I thought the cause of error was from `data` not being casted to `user`, so I imported the `user` and changed `data =>` to `(data: user) =>`. That didn't  work any better, because I received the same error.  

Am I correct `data` is what would be returned by the backend? The test example used results from Postman changed to the mouse user.

Any suggestions would be appreciated?

Deborah

**Comment by haxwell** [comment]
Hi Deborah.. 

So, first, the way I found out the problem here, is 
- by clicking the debug button in the main Karma test window (upper right)
- Going to the Developer Console in the resulting window (on my computer that's ctrl-alt-c)
- Going to the Console tab in Developer Console..

From there, you can see exceptions that don't appear in the command line view.

Anyhoo! With that said, the problem is that you are referring to `data` when actually, `data` refers to the object being sent. The first property in that object is also called data, so to access it, you would type `data.data`. This will clear up your immediate error. There's one more after that. You should catch it quick :)

Finally, what you are testing _in this case_ is that a `User` object is returned, but that will not always be the case. This test is for the `apiService`, and the `apiService` could be called upon to return all kinds of data. It would be better that in this test, you did your `expect()`s on a general set of data. So, in your `req.flush()` you would return, for example,

```
req.flush({foo: 1, bar: 2, baz: {id: 10, value: 'someValue'})
```
and in your tests you would test for 
```
expect(data.foo).toEqual(1)
expect(data.bar).toEqual(2)
expect(data.baz.id).toEqual(10)
```
and of course your subscribe would be like
```
....subscribe((data: Any) => {.....
```

**Comment by FrauDeborah** [comment]
Thank you Johnathan (@haxwell) for the lesson. It worked.  I'll add the other tests then push an update.

PS. I think you are psychic!  I had decided to try using the console in the Karma window. I added a `console.log( 'data = ' + JSON.stringify(data) )` and learned the data was being passed correctly from the req.flush(...). My next step was to try the  debug button. I had just logged into GitHub to paste a new screen print to updated status, only to fine your comments added 11 minutes earlier. Whoo hoo!

File: src/app/_services/api.service.spec.ts (hunk 1 above)
Comment (FrauDeborah) [reply src/app/_services/api.service.spec.ts]: fixed.

**Comment by FrauDeborah** [comment]
Hi Johnathan (@Haxwell),

I still have an error on the devApiTests branch. I committed changes and pushed so I could move to a different branch and come back to fix the error with the "tests apiService.get() for failure with deliberate error" later when I'm fresh. The good news is I'm getting a different error, but what ever is wrong also cause an error in the login dialog host tests. I checked login dialog host tests without the apiService.Get() failure test and the login dialog host tests passed. 

Hope I have the error fixed by tomorrow for the Sunday update.

Deborah

**Comment by FrauDeborah** [comment]
Hi Johnathan (@haxwell)

I finally got the tests for apiService.get() and apiServicepost() deliberate error tests to work. Yet, there is some sort of a transitory error somewhere within quizki tests. 

After I finally got the apiService.get() deliberate error test to work with no error, and found the login-dialog-host failed wanting the BrowserAnimationsModule added. I didn't add BrowserAnimationsModule because I wasn't working on login-dialog-host and I don't think its needed. When I commented out the apiService.get() deliberate error test, the login-dialog-host did not fail.

When I added the apiService.post() deliberate error test. The test worked, but the issue with login move from the login-dialog-host to the login component. Unlike the login-dialog-host, the login component test spec already imports BrowserAnimationsModule.

The actual error is the same for either login-dialog-host or the login component. Here is the screen capture for the login component.

image

Any Ideas? What would cause a transitory error like that?

Deborah

**Comment by haxwell** [comment]
@FrauDeborah No... but I know they sometimes happen with Jasmine/Karma/Frontend-testing-in-general.. I have never found a good answer as to why. :/

**Comment by FrauDeborah** [comment]
Hi  Johnathan,

Did you run ng test before merging? After I merged to local dev, I'm still getting the unrelated error on the LoginComponent. So I'm wondering if you got the error or if just happens in my environment.  Would not like it if my testing work should confuse other team members when they start writing tests. 

image

Deborah

**Comment by haxwell** [comment]
@FrauDeborah I did not run the tests beforehand; I should have. I get this same error, too. I think it is an actual error in the test, but I don't know exactly what the cause is yet. :/

**Pull request merged**